### PR TITLE
Terraform api-gateway module

### PR DESCRIPTION
**Pull request by oozywaters**

### CODE REVIEW

**Comment by oozywaters** [comment]
Ready for review

File: terraform/modules/api-gateway/main.tf [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+  # The /*/* portion grants access from any method on any resource
+  # within the API Gateway "REST API".
+  source_arn = "${aws_api_gateway_deployment.environment.execution_arn}/*/*"
+}
+
+resource "aws_api_gateway_account" "mxfactorial_api_account" {
+  cloudwatch_role_arn = "${aws_iam_role.cloudwatch.arn}"
+}
+
+resource "aws_iam_role" "cloudwatch" {
```
Comment (mxfactorial) [inline terraform/modules/api-gateway/main.tf]:
```suggestion
 resource "aws_iam_role" "api_gateway_cloudwatch" {
```

Comment (mxfactorial) [reply terraform/modules/api-gateway/main.tf]: conflicts avoided by specific identifiers

File: terraform/modules/api-gateway/main.tf [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        "Service": "apigateway.amazonaws.com"
+      },
+      "Action": "sts:AssumeRole"
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_iam_role_policy" "cloudwatch" {
```
Comment (mxfactorial) [inline terraform/modules/api-gateway/main.tf]:
```suggestion
 resource "aws_iam_role_policy" "api_gateway_cloudwatch" {
```

File: terraform/modules/api-gateway/main.tf [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  name = "${var.iam_role_policy_name}"
+  role = "${aws_iam_role.cloudwatch.id}"
```
Comment (mxfactorial) [inline terraform/modules/api-gateway/main.tf]:
```suggestion
   role = "${aws_iam_role.api_gateway_cloudwatch.id}"
```

File: terraform/modules/api-gateway/main.tf [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+  function_name = "${var.graphql_server_function_name}"
+  principal     = "apigateway.amazonaws.com"
+
+  # The /*/* portion grants access from any method on any resource
+  # within the API Gateway "REST API".
+  source_arn = "${aws_api_gateway_deployment.environment.execution_arn}/*/*"
+}
+
+resource "aws_api_gateway_account" "mxfactorial_api_account" {
+  cloudwatch_role_arn = "${aws_iam_role.cloudwatch.arn}"
```
Comment (mxfactorial) [inline terraform/modules/api-gateway/main.tf]:
```suggestion
   cloudwatch_role_arn = "${aws_iam_role.api_gateway_cloudwatch.arn}"
```

**Review by mxfactorial — CHANGES_REQUESTED** [review]
see notes

File: terraform/workspaces/api-gateway.tf [hunk 5]
```diff
@@ -1,7 +1,12 @@
 module "api_gateway" {
-  source = "git::https://github.com/systemaccounting/mxfactorial.git//terraform/modules/api-gateway?ref=terraform-api-gateway-module"
+  source = "../modules/api-gateway"
```
Comment (mxfactorial) [inline terraform/workspaces/api-gateway.tf:2]: :clap: :clap: :clap: :clap:

**Comment by mxfactorial** [comment]
specifically, error returned after applying in prod:
```
Error: Error applying plan:

1 error(s) occurred:

* module.api_gateway.aws_iam_role.cloudwatch: 1 error(s) occurred:

* aws_iam_role.cloudwatch: Error creating IAM Role mxfactorial-cloudwatch-role-prod: EntityAlreadyExists: Role with name mxfactorial-cloudwatch-role-prod already exists.
	status code: 409, request id: 57994dbb-08b8-11e9-a5fd-6defa18ece07
```

**Review by mxfactorial — APPROVED** [review]

**Pull request merged**